zips/zip-tbd1.md

```diff
@@ -0,0 +1,129 @@
+
+# Specification
+
+## Message URI syntax
+
+The current memo zip requires that the content of the message is less than 512 ASCII characters which is a sufficient size to include a URI which includes the sufficient details to locate and decrypt the message. Here is how the format of the pointer will be:
+
+## ABNF definition of message URI syntax
+    mmp-uri                     = mmp ":" version ":" cid query-params
+    mmp-uri                     = /1*512VCHAR; Limit to 512 ASCII VCHARs
+    scheme                      = "mmp"
+    version                       = 1*2DIGIT
+    message-location      = 1*256VCHAR
+    query-params            = "?" encrypt-key-param "&" param *("&" param)
+    param                        = param-name "=" param-value
+    param-name              = 1*ALPHA
+    param-value              = 1*VCHAR
+
+## Example of message URI
+
+mmp:v1:bafybeihdwdcefgh4dqkjv67uzcmw7ojee6xedzdetojuzjevtenxquvyku?ttl=2024-07-02T14:30:00Z&key=Hy9X_k2mLpQrZtNbVc5hA7sDxEuFoP-iQnWyG4M6OjBv
+
+## Protocol Versioning
+
+Since the version number determines the various associated properties of the protocol (including cryptographic properties to support cryptographic agility) there can be a wide range of capabilities that MAY NOT be interoperable. As such, all valid implementations of this protocol MUST implement at least version 1 to ensure a fallback for messaging. 
+
+## Message Location
+
+This message location is a URL which can be used to locate a message. It’s used to locate the encrypted ciphertext of the media which is stored at the URL location. It MUST be defined by each protocol version specification. For example in V1, this will be defined as a CID that can be used to get the message from IPFS, but in future iterations it could be a URL that points to a S3 bucket, Torrent file, or elsewhere. It MAY contain additional details about how a path as defined by [^RFC-3986] can be used if a version wants to support this. If the URL contains additional information such as a different scheme (e.g. `https://`) it MUST be defined how it’s encoded within the protocol versioning specification since this URI 
+
+## Query Parameters
+
+All query parameters are optional and MAY be included. Any required query parameters MUST be defined by the version specification. The exception here is the key parameter which are defined in this document because of its expected ubiquitous use across versions.
+
+### Encryption Key Query Parameter
+In all versions of the Media Memo Protocol that require passing a secret key between sender and recipients the key parameter MUST be used to. The key parameter MUST be the first query parameter if used so it should follow directly after a ? symbol. Reuse of this query parameter for many protocols will ensure the greatest possible interoperability.
+
+### ABNF for Encryption Key Query Parameter
+    encrypt-key-param = "key=" 1*64VCHAR
```

Comment:
This doesn't define the `ttl` field that you show in your example; under the current ABNF, it would need to be part of the `encrypt-key-param` production, which is required to appear first. Instead of defining an explicit initial `encrypt-key-param` production that must appear before other query parameter pairs, permit an arbitrary order and then just require the `key` and potentially `ttl` (if intended) keys be present in prose.

Reply:
The various options I considered here were:

1. Add an additional `:` delimited section after version, but before the message location that is the key, but this becomes weird if someone opts to not have a key all together in a version. This could be solved with an empty section that produces `::` if we want and saves a few bytes.
2. Reserve use of fragments for passing the key only and prevent usage of other fragments from being used
3. query parameter specific so that it can be dropped or added as needed.

The reason I opted to go with the query parameter option is that I expected that the parsers would handle this fine, but as soon as someone passed this to a URL resolver it would fail in which case it wouldn't make a network call at all. The one edge case that I can think of here is the registration of custom scheme handlers in operating systems and some browsers which may end up passing the URI around locally on device, but would likely still fail to resolve them in which case I couldn't see a reason that the query parameter didn't make sense.

However, in looking into the [behavior of fragments](https://en.wikipedia.org/wiki/URI_fragment#Examples) a bit more for different media types, it does look possible to define multiple such as the example for PDFs.

In this case, I'm convinced that using fragments is the better option here but will keep the syntax matching what's described (well with some modifications to the characters that can be used as the questions above around VCHAR apply here too).

Reply:
Also, I opted to just drop the ttl example from this spec. Instead it seems more relevant that this is defined in the actual v1 spec.